src/main/java/com/google/sps/servlets/EditEventServlet.java

```diff
@@ -0,0 +1,215 @@
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userToken = request.getParameter("userToken");
+    if (Firebase.isUserLoggedIn(userToken)) {
+      String userID = Firebase.authenticateUser(userToken);
+
+      Key keyRequested = getEventKey(request, "key");
+      Query query = new Query("Event", keyRequested);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Entity eventRequested = datastore.prepare(query).asSingleEntity();
+
+      Entity eventEntity = updateEvent(request, eventRequested);
+      datastore.put(eventEntity);
+    } else {
+      throw new IOException("Cannot edit an event while not logged in");
+    }
+
+    // Redirect back to the my-events HTML page.
+    response.sendRedirect("/my-events.html");
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws ServletException, IOException {
+    Key keyRequested;
+    try {
+      keyRequested = getEventKey(request, "Event");
+      if (keyRequested != null) {
+        Query query = new Query("Event", keyRequested);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity eventRequested = datastore.prepare(query).asSingleEntity();
+
+        String userToken = request.getParameter("userToken");
+        if (userToken != null) {
+          if (Firebase.isUserLoggedIn(userToken)) {
+            String userID = Firebase.authenticateUser(userToken);
+            Key userKey = KeyFactory.createKey("User", userID);
+            try {
+              Entity userEntity = datastore.get(userKey);
+
+              if (userEntity != null) {
+                String creator = eventRequested.getProperty("creator").toString();
+
+                if (creator.equals(userID)) {
+                  request = populateRequest(request, eventRequested);
+                  request
+                      .getRequestDispatcher("/WEB-INF/jsp/edit-event-form.jsp")
+                      .forward(request, response);
+                } else {
+                  request
+                      .getRequestDispatcher("/WEB-INF/jsp/access-denied.jsp")
+                      .forward(request, response);
+                }
+              }
+            } catch (EntityNotFoundException exception) {
+              // datastore entry has not been created yet for this user, create it now
+              Entity entity = new Entity(userKey);
+              entity.setProperty("firebaseID", userID);
+              datastore.put(entity);
+              request
+                  .getRequestDispatcher("/WEB-INF/jsp/access-denied.jsp")
+                  .forward(request, response);
+            }
+          }
+        }
+      }
+    } catch (IllegalArgumentException | IOException | NullPointerException | ServletException e) {
```

Comment:
It is not immediately clear to the reader which parts of the code throw these exceptions. Maybe handle them where they are thrown (like the EntityNotFoundException above), or add some comments to document where these exceptions originate.

Reply:
Will do!